File: Server/app/services/watchlist_service.py
from __future__ import annotations

import secrets

from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession


class WatchlistService:
    def __init__(self, db: AsyncSession) -> None:
        self.db = db

    async def get_all_default_watchlist_symbols(self) -> list[str]:
        result = await self.db.execute(
            text(
                """
                SELECT DISTINCT watchlist_items.symbol
                FROM watchlists
                JOIN watchlist_items ON watchlist_items.watchlist_id = watchlists.id
                WHERE watchlists.is_default = true
                ORDER BY watchlist_items.symbol ASC
                """
            )
        )
        return [r[0] for r in result.all()]

    async def get_default_watchlist(self, user_id: str) -> list[str]:
        result = await self.db.execute(
            text(
                """
                SELECT watchlist_items.symbol
                FROM watchlists
                JOIN watchlist_items ON watchlist_items.watchlist_id = watchlists.id
                WHERE watchlists.user_id = :user_id
                  AND watchlists.is_default = true
                ORDER BY watchlist_items.sort_order ASC
                """
            ),
            {"user_id": user_id},
        )
        return [r[0] for r in result.all()]

    async def set_default_watchlist(self, user_id: str, symbols: list[str]) -> list[str]:
        normalized = list(
            dict.fromkeys(
                s.strip().upper()
                for s in symbols
                if s.strip()
            )
        )

        if normalized:
            result = await self.db.execute(
                text("SELECT symbol FROM instruments WHERE symbol = ANY(:symbols)"),
                {"symbols": normalized},
            )
            valid = {r[0] for r in result.all()}
            filtered = [s for s in normalized if s in valid]
        else:
            filtered = []

        result = await self.db.execute(
            text(
                """
                SELECT id FROM watchlists
                WHERE user_id = :user_id AND is_default = true
                LIMIT 1
                """
            ),
            {"user_id": user_id},
        )
        row = result.one_or_none()
        watchlist_id = row[0] if row else secrets.token_hex(16)

        await self.db.execute(
            text(
                """
                INSERT INTO watchlists (id, user_id, name, is_default)
                VALUES (:id, :user_id, 'Default', true)
                ON CONFLICT (user_id, name) DO UPDATE SET is_default = true
                """
            ),
            {"id": watchlist_id, "user_id": user_id},
        )

        await self.db.execute(
            text("DELETE FROM watchlist_items WHERE watchlist_id = :wid"),
            {"wid": watchlist_id},
        )

        for idx, symbol in enumerate(filtered):
            await self.db.execute(
                text(
                    """
                    INSERT INTO watchlist_items (id, watchlist_id, symbol, sort_order)
                    VALUES (:id, :wid, :symbol, :order)
                    """
                ),
                {
                    "id": secrets.token_hex(16),
                    "wid": watchlist_id,
                    "symbol": symbol,
                    "order": idx,
                },
            )

        return filtered

File: Server/src/watchlists/watchlistService.ts
import crypto from 'node:crypto'
import type { Pool } from 'pg'

export class WatchlistService {
  constructor(private readonly pool: Pool) {}

  async getAllDefaultWatchlistSymbols() {
    const result = await this.pool.query<{ symbol: string }>(
      `
        select distinct watchlist_items.symbol
        from watchlists
        join watchlist_items on watchlist_items.watchlist_id = watchlists.id
        where watchlists.is_default = true
        order by watchlist_items.symbol asc
      `,
    )

    return result.rows.map((row) => row.symbol)
  }

  async getDefaultWatchlist(userId: string) {
    const result = await this.pool.query<{ symbol: string }>(
      `
        select watchlist_items.symbol
        from watchlists
        join watchlist_items on watchlist_items.watchlist_id = watchlists.id
        where watchlists.user_id = $1
          and watchlists.is_default = true
        order by watchlist_items.sort_order asc
      `,
      [userId],
    )

    return result.rows.map((row) => row.symbol)
  }

  async setDefaultWatchlist(userId: string, symbols: string[]) {
    const normalized = [...new Set(symbols.map((symbol) => symbol.trim().toUpperCase()).filter(Boolean))]
    const instrumentRows = await this.pool.query<{ symbol: string }>(
      'select symbol from instruments where symbol = any($1::text[])',
      [normalized],
    )

    const validSymbols = new Set(instrumentRows.rows.map((row) => row.symbol))
    const filtered = normalized.filter((symbol) => validSymbols.has(symbol))

    const existing = await this.pool.query<{ id: string }>(
      `
        select id from watchlists
        where user_id = $1 and is_default = true
        limit 1
      `,
      [userId],
    )

    const watchlistId = existing.rows[0]?.id ?? crypto.randomUUID()
    await this.pool.query(
      `
        insert into watchlists (id, user_id, name, is_default)
        values ($1, $2, 'Default', true)
        on conflict (user_id, name)
        do update set is_default = true
      `,
      [watchlistId, userId],
    )

    await this.pool.query('delete from watchlist_items where watchlist_id = $1', [watchlistId])

    for (const [index, symbol] of filtered.entries()) {
      await this.pool.query(
        `
          insert into watchlist_items (id, watchlist_id, symbol, sort_order)
          values ($1, $2, $3, $4)
        `,
        [crypto.randomUUID(), watchlistId, symbol, index],
      )
    }

    return filtered
  }
}
